pkg/reconciler/workload/apply.go

```diff
@@ -40,52 +36,41 @@ var (
 	errPatchKubeApp          = "cannot patch existing KubernetesApplication"
 )
 
-// KubeAppApply applies changes to the supplied KubernetesApplication. The
-// KubernetesApplication will be created if it does not exist, or patched if it
-// does. Resource templates are merged instead of replaced if they have the same
-// name and GroupVersionKind.
-func KubeAppApply(ctx context.Context, c client.Client, o runtime.Object, ao ...resource.ApplyOption) error {
-	opts := &resource.ApplyOptions{}
-	for _, fn := range ao {
-		fn(opts)
-	}
-
-	m, ok := o.(*workloadv1alpha1.KubernetesApplication)
-	if !ok {
-		return errors.New(errNotKubeApp)
-	}
-
-	desired := &workloadv1alpha1.KubernetesApplication{}
-	m.DeepCopyInto(desired)
-
-	err := c.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, o)
-	if kerrors.IsNotFound(err) {
-		return errors.Wrap(c.Create(ctx, o), errCreateKubeApp)
-	}
-	if err != nil {
-		return errors.Wrap(err, errGetKubeApp)
-	}
-
-	if opts.ControllersMustMatch && !meta.HaveSameController(m, desired) {
-		return errors.New(errDiffController)
-	}
-
-	for i, temp := range m.Spec.ResourceTemplates {
-		if desired.Spec.ResourceTemplates[i].Spec.Template.GroupVersionKind() != temp.Spec.Template.GroupVersionKind() {
-			continue
+// KubeAppApplyOption ensures resource templates are merged instead of replaced
+// before patch if they have the same name and GroupVersionKind.
+func KubeAppApplyOption() resource.ApplyOption {
+	return func(_ context.Context, current, desired runtime.Object) error {
+		c, ok := current.(*workloadv1alpha1.KubernetesApplication)
+		if !ok {
+			return errors.New(errNotKubeApp)
 		}
-		if desired.Spec.ResourceTemplates[i].GetName() != temp.GetName() {
-			continue
+		d, ok := desired.(*workloadv1alpha1.KubernetesApplication)
+		if !ok {
+			return errors.New(errNotKubeApp)
 		}
-		if err := mergo.Merge(desired.Spec.ResourceTemplates[i].Spec.Template, temp.Spec.Template); err != nil {
-			return errors.Wrap(err, errMergeKubeAppTemplates)
+		for i, temp := range c.Spec.ResourceTemplates {
+			if d.Spec.ResourceTemplates[i].Spec.Template.GroupVersionKind() != temp.Spec.Template.GroupVersionKind() {
+				continue
+			}
+			if d.Spec.ResourceTemplates[i].GetName() != temp.GetName() {
+				continue
+			}
+			jc, err := json.Marshal(temp.Spec.Template)
+			if err != nil {
+				return errors.Wrap(err, errMergeKubeAppTemplates)
+			}
+			jd, err := json.Marshal(d.Spec.ResourceTemplates[i].Spec.Template)
+			if err != nil {
+				return errors.Wrap(err, errMergeKubeAppTemplates)
+			}
+			merge, err := jsonpatch.MergePatch(jc, jd)
```

Comment:
Nit: Maybe named this `merged`? I confused `jsonpatch.MergePatch` with `jsonpatch.CreateMergePatch` and thought `merge` was a merge patch, not the product of applying a merge patch until I did some reading.